packages/metadata/src/decorator-factory.ts

```diff
@@ -0,0 +1,539 @@
+    methodName?: string | symbol,
+    methodDescriptor?: TypedPropertyDescriptor<any> | number,
+  ) {
+    if (localMeta == null) localMeta = {};
+    if (localMeta[methodName!] != null) {
+      throw new Error(
+        'Decorator cannot be applied more than once on ' +
+          this.getTargetName(target, methodName, methodDescriptor),
+      );
+    }
+    // Set the method metadata
+    localMeta[methodName!] = this.withTarget(this.spec, target);
+    return localMeta;
+  }
+
+  create(): MethodDecorator {
+    return (
+      target: Object,
+      methodName: string | symbol,
+      descriptor: TypedPropertyDescriptor<any>,
+    ) => this.decorate(target, methodName, descriptor);
+  }
+
+  /**
+   * Create a method decorator function
+   * @param key Metadata key
+   * @param spec Metadata object from the decorator function
+   * @param options Options for the decorator
+   */
+  static createDecorator<T>(key: string, spec: T, options?: DecoratorOptions) {
+    return super._createDecorator<T, MetadataMap<T>, MethodDecorator>(
+      key,
+      spec,
+      options,
+    );
+  }
+}
+
+/**
+ * Factory for parameter decorators
+ */
+export class ParameterDecoratorFactory<T> extends DecoratorFactory<
+  T,
+  MetadataMap<T[]>,
+  ParameterDecorator
+> {
+  protected processInherited(
+    baseMeta: MetadataMap<T[]>,
+    target: Object,
+    methodName?: string | symbol,
+    parameterIndex?: TypedPropertyDescriptor<any> | number,
+  ) {
+    const method = methodName ? methodName : '';
+    let methodMeta = baseMeta[method];
+    if (methodMeta == null) {
+      // Initialize the method metadata
+      methodMeta = new Array(
+        this.getNumberOfParameters(target, methodName),
+      ).fill(undefined);
+      baseMeta[method] = methodMeta;
+    }
+    const index = parameterIndex as number;
+    methodMeta[index] = this.withTarget(
+      <T>this.inherit(methodMeta[index]),
+      target,
+    );
+    return baseMeta;
+  }
+
+  protected processLocal(
+    localMeta: MetadataMap<T[]>,
+    target: Object,
+    methodName?: string | symbol,
+    parameterIndex?: TypedPropertyDescriptor<any> | number,
+  ) {
+    const method = methodName ? methodName : '';
+    if (localMeta == null) localMeta = {};
+    // Find the method metadata
+    let methodMeta = localMeta[method];
+    if (methodMeta == null) {
+      // Initialize the method metadata
+      methodMeta = new Array(
+        this.getNumberOfParameters(target, methodName),
+      ).fill(undefined);
+      localMeta[method] = methodMeta;
+    }
+    const index = parameterIndex as number;
+    if (this.getTarget(methodMeta[index]) === target) {
+      throw new Error(
+        'Decorator cannot be applied more than once on ' +
+          this.getTargetName(target, methodName, parameterIndex),
+      );
+    }
+    // Set the parameter metadata
+    methodMeta[index] = this.withTarget(
+      <T>this.inherit(methodMeta[index]),
+      target,
+    );
+    return localMeta;
```

Comment:
These two methods (`processInherited` and `processLocal`) look too similar to me, it's hard to tell what are the important differences and what is the shared boiler-plate. Could you please try another round of extracting shared code to clean this up?

For example, the following block of code can be easily extracted into a shared helper method.

```ts
// processInherited
    let methodMeta = baseMeta[method];
    if (methodMeta == null) {
      // Initialize the method metadata
      methodMeta = new Array(
        this.getNumberOfParameters(target, methodName),
      ).fill(undefined);
      baseMeta[method] = methodMeta;
    }

// processLocal
    let methodMeta = localMeta[method];
    if (methodMeta == null) {
      // Initialize the method metadata
      methodMeta = new Array(
        this.getNumberOfParameters(target, methodName),
      ).fill(undefined);
      localMeta[method] = methodMeta;
    }
```

With my proposed helper:

```ts
// processInherited
  const methodMeta = getOrCreateMeta(baseMeta, target, method);

// processLocal
  const methodMeta = getOrCreateMeta(localMeta, target, method);

// getOrCreateMeta(meta, target, method)
    let methodMeta = meta[method];   
    if (methodMeta == null) {
      // Initialize the method metadata
      methodMeta = new Array(
        this.getNumberOfParameters(target, method || ''),
      ).fill(undefined);
      meta[method] = methodMeta;
    }
    return methodMeta;
```